### PR TITLE
Fix `lr_scheduler` unexpectedly calls `step()` when init argument last_epoch is larger than -1

### LINKED ISSUE #102261: ExponentialLR unexpectedly calls `step()` when init argument `last_epoch` is larger than -1
### 🐛 Describe the bug

Currently, the init function of ``torch.optim.lr_scheduler._LRSchedule`` will call ``self.step()`` once. This causes a mismatch between the learning rate used by the optimizer and the closed_form_lr of ExponentialLR, when init argument `last_epoch` is larger than -1.

```python
import torch

model = torch.nn.Linear(3, 3)
optim = torch.optim.AdamW(model.parameters())
sched = torch.optim.lr_scheduler.ExponentialLR(optim, gamma=0.999)

optim.step()
sched.step()
print("Optim & sched:")
print(optim.state_dict())
print(optim.param_groups[0]["lr"])
print(sched.state_dict())
print(sched._get_closed_form_lr())
print("")

# As if we are restoring from a checkpoint
optim2 = torch.optim.AdamW(model.parameters())
optim2.load_state_dict(optim.state_dict())
# Init scheduler with last_epoch=0
sched2 = torch.optim.lr_scheduler.ExponentialLR(optim2, gamma=0.999, last_epoch=0)

print("Optim2 & sched2:")
print(optim2.state_dict())
print(optim2.param_groups[0]["lr"])
print(sched2.state_dict())
print(sched2._get_closed_form_lr())
print("")

```

As the result shows, ``optim2`` has lr 0.000998001, but the closed form lr of ``sched2`` is 0.000999. This behavior causes confusion and inconsistency when one resumes training from a checkpoint.

### Versions

Collecting environment information...
PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

### PR DESCRIPTION
**Pull request by zeshengzong**
Fixes #102261

## Changes

- Use flag `_is_initial` to replace `self.last_epoch == 0` condition to judge whether `lr` should be initial value
- Add test for `ExponentialLR` checkpoint usecase

## Test Result

```python
pytest -s test/optim/test_lrscheduler.py  -vv
```

image

### CODE REVIEW

**Comment by zeshengzong** [comment]
Hello @albanD @janeyx99 , please check whether the fixing is feasible, if it works, I would like to continue fix more schedulers which have same problem, like `MultiplicativeLR`, `LinearLR`, thanks!

**Review by janeyx99 — DISMISSED** [review]
This does not look like the right approach. If the discrepancy is for ExponentialLR between get_lr and _get_closed_form_lr, I'd expect the fix to be local there. Could you explain your approach a little bit?

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -724,7 +738,7 @@ def get_lr(self):
         """Compute the learning rate of each parameter group."""
         _warn_get_lr_called_within_step(self)
 
-        if self.last_epoch == 0:
+        if self._is_initial:
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:741]:
```suggestion
        // when loading from a checkpoint, we don't want _initial_step (called from the constructor) to update the lr
        // one more step ahead of itself.
        if self._is_initial:
```

**Review by janeyx99 — COMMENTED** [review]
Oh actually, I see what you're doing now. Sorry I was confused yesterday. I'm willing to accept this fix if you update the test case.

It would also be good to include a comment about why we prefer the `_is_initial`.

**Comment by joecummings** [comment]
I'd love to see this expanded to ensure this works for all LRSchedulers! I have confirmed that I see the same issue when testing with StepLR (when I try to resume training and setup a new LRScheduler, it is always one step off b/c of this initial step that is taken in the init of LRSchedulers).

**Comment by janeyx99** [comment]
@zeshengzong lmk if you can bring this PR over the finish line with expanding it to all LRSchedulers!

**Comment by zeshengzong** [comment]
> @zeshengzong lmk if you can bring this PR over the finish line with expanding it to all LRSchedulers!

Hi @janeyx99 , sorry for late reply, busy with something else previously. I would like fix all of them and hope I could clean up all issues relatvie with `lr_scheduler`, thanks for help!

**Comment by zeshengzong** [comment]
> Oh actually, I see what you're doing now. Sorry I was confused yesterday. I'm willing to accept this fix if you update the test case.
> 
> It would also be good to include a comment about why we prefer the `_is_initial`.

Yes, adding a context to better distinguish initial lr or calculate lr, `self.last_epoch == 0` is not enough at this case.

File: test/optim/test_lrscheduler.py [hunk 2]
```diff
@@ -2558,6 +2558,23 @@ def test_constant_initial_params_swalr(self):
...
+    def test_lr_scheduler_checkpoint(self, LRClass):
+        model = torch.nn.Linear(3, 3)
+        optim = torch.optim.AdamW(model.parameters())
+        sch = LRClass(optim)
+        optim.step()
+        sch.step()
+        optim2 = torch.optim.AdamW(model.parameters())
+        optim2.load_state_dict(optim.state_dict())
+        sch2 = LRClass(optim2, last_epoch=1)
+        self.assertEqual(optim.param_groups[0]["lr"], optim2.param_groups[0]["lr"])
```
[the remark this answers is not in the document]
Comment (zeshengzong) [reply test/optim/test_lrscheduler.py]: Changed, thanks!

File: torch/optim/lr_scheduler.py [hunk 3]
```diff
@@ -134,7 +135,8 @@ def wrapper(*args, **kwargs):
     def _initial_step(self):
         """Initialize step counts and perform a step."""
```
[the remark this answers is not in the document]
Comment (zeshengzong) [reply torch/optim/lr_scheduler.py:136]: I think one of the key effect is to initialize optimizer initial lr as the same as the scheduler lr when create it, and reuse this part of code:

https://github.com/pytorch/pytorch/blob/4c5cf18ee0503fffa910c70a63a054bfbc55a480/torch/optim/lr_scheduler.py#L200-L220

One improvement can be made is extracting internal update lr logic from `step` public method, please check this PR: https://github.com/pytorch/pytorch/pull/149392 and the issue it fixed. Thanks!

File: test/optim/test_lrscheduler.py [hunk 4]
```diff
@@ -2558,6 +2558,23 @@ def test_constant_initial_params_swalr(self):
                 self.assertEqual(group["swa_lr"], 0.05)
                 self.assertEqual(sch.base_lrs, [0.1])
 
+    @parametrize(
+        "LRClass",
+        [
+            partial(ExponentialLR, gamma=0.999),
+        ],
+    )
```
Comment (janeyx99) [inline test/optim/test_lrscheduler.py:2566]: It'd be great to expand this to more than ExponentialLR!

Comment (zeshengzong) [reply test/optim/test_lrscheduler.py:2566]: Participating a pytorch meetup, will do it next week, thanks! :D

Comment (zeshengzong) [reply test/optim/test_lrscheduler.py:2595]: Hi @janeyx99 , I've added more schedulers in here, `ReduceLROnPlateau` has different pattern, so I separate it to another test.

File: test/optim/test_lrscheduler.py [hunk 5]
```diff
@@ -2575,6 +2575,46 @@ def test_constant_initial_params_swalr(self):
...
+    def test_lr_scheduler_checkpoint(self, LRClass):
+        model = torch.nn.Linear(3, 3)
+        optim = torch.optim.AdamW(model.parameters())
+        sch = LRClass(optim)
+        optim.step()
+        sch.step()
+        optim2 = torch.optim.AdamW(model.parameters())
+        optim2.load_state_dict(optim.state_dict())
+        sch2 = LRClass(optim2, last_epoch=0)
+        self.assertEqual(sch2.get_last_lr()[0], optim.param_groups[0]["lr"])
```
Comment (zeshengzong) [inline test/optim/test_lrscheduler.py:2605]: Replaced with `get_last_lr` since some schedulers not implemented `_get_closed_form_lr` method.

Comment (janeyx99) [reply test/optim/test_lrscheduler.py:2605]: Can we use _get_closed_form_lr whenever it is available (using hasattr)

**Review by janeyx99 — APPROVED** [review]
